drivers/hall_effect/hall_effect.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * SPDX-FileCopyrightText: 2025 TU Dresden
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+/**
+ * @ingroup     drivers_hall_effect
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for a generic hall effect sensor
+ *
+ * @author      Leonard Herbst <leonard.herbst@tu-dresden.de>
+ *
+ * @}
+ */
+
+#include "hall_effect.h"
+#include "hall_effect_params.h"
+
+#include <errno.h>
+#include "log.h"
+#include "ztimer.h"
+#include "time_units.h"
+
+/**
+ * @brief Scaling factor to apply to adjust for the gear reduction ratio being in tenths.
+ */
+#define GEAR_RED_RATIO_SCALE 10
+
+/* Prototypes */
+
+static void _pulse_callback(void *arg);
+static void _read_reset_pulse_counter(hall_effect_t *dev, int32_t *pulse_counter);
+static bool _read_delta_t_direction(hall_effect_t *dev, uint32_t *delta_t, bool *ccw);
+
+/* Public API */
+
+int hall_effect_init(hall_effect_t *dev, const hall_effect_params_t *params)
+{
+    dev->params = *params;
+
+    if (gpio_init(dev->params.direction, GPIO_IN)) {
+        LOG_ERROR("[hall_effect] Failed configuring the direction pin as an input!\n");
+        return -EIO;
+    }
+
+    dev->delta_t = 0;
+    dev->ccw = false;
+    dev->stale = false;
+    dev->pulse_counter = 0;
+    dev->last_read_time = ztimer_now(ZTIMER_USEC);
+
+    if (gpio_init_int(dev->params.interrupt, GPIO_IN, GPIO_RISING, _pulse_callback, (void *) dev)) {
+        LOG_ERROR("[hall_effect] Failed configuring the interrupt pin!\n");
+        return -EIO;
+    }
+
+    return 0;
+}
+
+int hall_effect_read_rpm(hall_effect_t *dev, int32_t *rpm)
+{
+    uint32_t delta_t;
+    bool ccw;
+    if (!_read_delta_t_direction(dev, &delta_t, &ccw)) {
+        *rpm = 0;
+        return 0;
+    }
+
+    /* delta_t represents the number of micro seconds since the last pulse.
+     * Invert and divide by the number of micro seconds per minute
+     * to obtain the rpm. Apply scaling factors like gear reduction
+     * or pulses per revolution.
+     */
+    *rpm = SEC_PER_MIN * US_PER_SEC * GEAR_RED_RATIO_SCALE
+           / (delta_t * CONFIG_HALL_EFFECT_PPR * CONFIG_HALL_EFFECT_GEAR_RED_RATIO);
+    if (ccw) {
+        *rpm *= -1;
+    }
+    return 0;
+}
+
+int hall_effect_read_reset_ceti_revs(hall_effect_t *dev, int32_t *pulse_counter)
+{
+    _read_reset_pulse_counter(dev, pulse_counter);
+    *pulse_counter *= 100 * GEAR_RED_RATIO_SCALE;
+    *pulse_counter /= CONFIG_HALL_EFFECT_PPR;
+    *pulse_counter /= CONFIG_HALL_EFFECT_GEAR_RED_RATIO;
+    return 0;
+}
+
+/* Private API */
+
+/* Triggered on the high flank of a pulse */
+static void _pulse_callback(void *arg)
+{
+    hall_effect_t *dev = (hall_effect_t *) arg;
+
+    uint32_t now = ztimer_now(ZTIMER_USEC);
+
+    /* Reading the shifted phase: low -> cw, high -> ccw */
+    dev->ccw = gpio_read(dev->params.direction);
+    if (now < dev->last_read_time) {
+        /* timer had an overflow */
+        dev->delta_t = UINT32_MAX - dev->last_read_time + now + 1;
+    } else {
```

Comment:
```suggestion
    }
    else {
```

As per our coding convention: https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#indentation-and-braces